### PR TITLE
Traffic-Based App Rewards: Fix hashing of batches

### LINKED ISSUE #4749: Use Splice.Amulet.CryptoHash functions directly once there available via rebase or merge
Affects  https://github.com/hyperledger-labs/splice/tree/adetokunbo/feature-cip-104-scan-computation
As of https://github.com/hyperledger-labs/splice/pull/4662` which adds db tests for these functions, they are inlined in the daml template that tests them
The inlined versions should be removed before the branch gets merged

### PR DESCRIPTION
**Pull request by adetokunbo**
Supercedes ~~Follow-up/Stacked on to~~ #5118
Fixes #4749

Fixes a bug in computation of hash of batches, and adds tests to confirm the fix

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a logical synchronizer upgrade test is required (from canton-3.5), comment `/lsu_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/hyperledger-labs/splice/blob/main/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/hyperledger-labs/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/CryptoHashEquivalenceIntegrationTest.scala [hunk 1]
```diff
@@ -212,62 +381,113 @@ object CryptoHashEquivalenceIntegrationTest {
...
+    case HashBatch(_) =>
+      throw new UnsupportedOperationException("HashBatch is not used in SQL expression tests")
   }
 
   // -- Test case definitions (fully static) -----------------------------------
 
+  // Decimal values to test format equivalence between Daml's show on Decimal
+  // and Postgres daml_numeric_to_text. Covers trailing zeros, negatives,
+  // small fractions, and integers.
+  val decimalFormatCases: Seq[String] = Seq(
```
Comment (dfordivam) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/CryptoHashEquivalenceIntegrationTest.scala]: How was this obtained, could you describe the source, such that us+ future readers know / verify the coverage

Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/CryptoHashEquivalenceIntegrationTest.scala]: From an brief look I noticed that max/min values are missing. so this must be hand-written.
Lets make sure we have full coverage / edge cases

File: daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml [hunk 2]
```diff
@@ -65,3 +66,16 @@ template CryptoHashProxy with
         childHashes : [Text]
       controller owner
       do pure (hashBatchOfBatches (map Hash childHashes)).value
+
+    nonconsuming choice CryptoHashProxy_HashDecimal : Text
+      with
+        input : Decimal
+      controller owner
+      do pure (hash (show input)).value
```
Comment (dfordivam) [inline daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml]: is `show` needed? we don't have instance for `Decimal`?

Comment (dfordivam) [reply daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml]: We use the instance in our code, please remove `show`
```
instance CryptoHash.Hashable MintingAllowance where
  hash MintingAllowance {provider, amount} =
    CryptoHash.hashRecord [CryptoHash.hash provider , CryptoHash.hash amount]
```

Comment (adetokunbo) [reply daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml]: Fixed

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/CryptoHashEquivalenceIntegrationTest.scala (hunk 1 above)
Comment (adetokunbo) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/CryptoHashEquivalenceIntegrationTest.scala]: Fixed

**Review by meiersi-da — APPROVED** [review]
Thanks.

**Comment by dfordivam** [comment]
@adetokunbo Now that the other daml PR is closed, why don't you rebase this directly on feature-branch so that this could be merged. Please ensure CI is green before merge

**Comment by adetokunbo** [comment]
@meiersi-da: PTAL, only the relevant commits are present.

File: daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml [hunk 3]
```diff
@@ -58,36 +18,49 @@ template CryptoHashProxy with
...
       controller owner
-      do pure (hashMintingAllowance provider amount)
+      do pure (hash (MintingAllowance provider amount)).value
 
     nonconsuming choice CryptoHashProxy_HashBatchOfMintingAllowances : Text
       with
         allowanceHashes : [Text]
       controller owner
-      do pure (hashBatchOfMintingAllowances allowanceHashes)
+      do pure (hashVariant "BatchOfMintingAllowances" [hash (map Hash allowanceHashes)]).value
```
Comment (meiersi-da) [inline daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml:47]: It would be nicer if we called the actual function and not a copy. That would require passing in the allowances directly. It's OK to merge as-is though, as we would find a problem in our e2e integration test for reward accounting.

File: daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml [hunk 4]
```diff
@@ -58,36 +18,49 @@ template CryptoHashProxy with
...
 
     nonconsuming choice CryptoHashProxy_HashBatchOfBatches : Text
       with
         childHashes : [Text]
       controller owner
-      do pure (hashBatchOfBatches childHashes)
+      do pure (hashVariant "BatchOfBatches" [hash (map Hash childHashes)]).value
```
Comment (meiersi-da) [inline daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml:53]: Same here

**Review by meiersi-da — APPROVED** [review]
Thanks.

File: daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml (hunk 3 above)
Comment (dfordivam) [reply daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml:47]: Doing `hash` on actual inputs is already part of `CryptoHashProxy_HashBatch`, which includes allowances. So these existing (older) APIs are redundant in that sense, (unless they are are testing edge cases not covered already by `CryptoHashProxy_HashBatch`)

Comment (adetokunbo) [reply daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml:47]: > It's OK to merge as-is though
Thank you - I'll do that

File: daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml (hunk 4 above)
Comment (adetokunbo) [reply daml/splice-amulet-test/daml/Splice/Testing/CryptoHashProxy.daml:53]: Noted, as before, I'll merge as-is
